File: app/src/main/java/com/example/agriventure/ui/market/BuyerOffersFragment.java
package com.example.agriventure.ui.market;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.ProgressBar;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatImageView;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.agriventure.R;
import com.example.agriventure.data.adapter.BuyerOffersAdapter;
import com.example.agriventure.data.models.Offer;
import com.example.agriventure.ui.BaseFragment;
import com.example.agriventure.util.Constants;
import com.google.android.material.card.MaterialCardView;
import com.google.android.material.textview.MaterialTextView;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

import java.util.ArrayList;
import java.util.List;

public class BuyerOffersFragment extends BaseFragment {

    private List<Offer> allOfferList;
    private RecyclerView allOffersRv;
    private MaterialTextView marketEmptyText, acceptedText, pendingText, declinedText;
    private LinearLayout statsHolder;
    private ProgressBar mProgressBar;
    private AppCompatImageView emptyImage;
    private MaterialCardView cardView;

    public View onCreateView(@NonNull LayoutInflater inflater,
                             ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_offers, container, false);
        
        allOffersRv = view.findViewById(R.id.all_market_products);

        marketEmptyText = view.findViewById(R.id.offers_empty);
        statsHolder = view.findViewById(R.id.stats_holder);
        acceptedText = view.findViewById(R.id.total_accepted);
        pendingText = view.findViewById(R.id.total_pending);
        declinedText = view.findViewById(R.id.total_declined);

        mProgressBar = view.findViewById(R.id.data_loading);

        emptyImage = view.findViewById(R.id.img_offers_empty);

        cardView = view.findViewById(R.id.item_card);

        allOfferList = new ArrayList<>();
        return view;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        DatabaseReference mDatabase = FirebaseDatabase.getInstance().getReference().child("offers");

        mDatabase.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot snapshot) {
                allOfferList.clear();
                for (DataSnapshot dataSnapshot : snapshot.getChildren()){
                    Offer offer = dataSnapshot.getValue(Offer.class);
                    if (offer.getBuyer_name().equals(Constants.buyerName)){
                        allOfferList.add(offer);
                    }
                }
                updateView();
            }

            @Override
            public void onCancelled(@NonNull DatabaseError error) {

            }
        });

    }

    private void updateView() {
        if (allOfferList.size()>0){
            statsHolder.setVisibility(View.VISIBLE);
            setUpMyOffers(allOfferList);
        }else{
            marketEmptyText.setVisibility(View.VISIBLE);
            emptyImage.setVisibility(View.VISIBLE);
            mProgressBar.setVisibility(View.GONE);
        }
    }

    private void setUpMyOffers(List<Offer> offers) {
        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(activity, RecyclerView.VERTICAL, false);
        BuyerOffersAdapter offersAdapter = new BuyerOffersAdapter(activity, offers);
        setUpStats(offers);
        allOffersRv.setAdapter(offersAdapter);
        allOffersRv.setLayoutManager(linearLayoutManager);
        allOffersRv.setVisibility(View.VISIBLE);
        allOffersRv.setVisibility(View.VISIBLE);
        cardView.setVisibility(View.VISIBLE);
        mProgressBar.setVisibility(View.GONE);
    }

    private void setUpStats(List<Offer> offers) {
        int accepted = 0;
        int pending = 0;
        int declined = 0;
        for (int i=0;i<offers.size();i++){
            Offer o = offers.get(i);
            if (o.getOffer_status().equals("Accepted")){
                accepted+=1;
            }else if (o.getOffer_status().equals("Pending")){
                pending+=1;
            }else{
                declined+=1;
            }
        }
        acceptedText.setText(String.valueOf(accepted));
        pendingText.setText(String.valueOf(pending));
        declinedText.setText(String.valueOf(declined));
    }
}

File: app/src/main/java/com/example/agriventure/ui/market/BuyerOffersFragment.kt
package com.example.agriventure.ui.market

import android.content.Context
import com.example.agriventure.ui.BaseFragment
import com.example.agriventure.data.models.Offer
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.textview.MaterialTextView
import android.widget.LinearLayout
import android.widget.ProgressBar
import androidx.appcompat.widget.AppCompatImageView
import com.google.android.material.card.MaterialCardView
import android.view.LayoutInflater
import android.view.ViewGroup
import android.os.Bundle
import android.view.View
import com.example.agriventure.R
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ValueEventListener
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.agriventure.data.adapter.BuyerOffersAdapter
import com.example.agriventure.databinding.FragmentOffersBinding
import com.example.agriventure.util.Constants
import com.google.firebase.database.ktx.database
import com.google.firebase.ktx.Firebase
import java.util.ArrayList

class BuyerOffersFragment : BaseFragment() {
    private lateinit var allOfferList: MutableList<Offer>
    private lateinit var binding : FragmentOffersBinding
    private lateinit var businessName : String
    private val databaseRef : DatabaseReference by lazy{
        Firebase.database.reference.child("offers")
    }
    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?, savedInstanceState: Bundle?
    ): View? {
        binding = FragmentOffersBinding.inflate(inflater, container, false)
        allOfferList = ArrayList()
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        businessName =
            activity.getPreferences(Context.MODE_PRIVATE).getString(Constants.buyerBusinessName, "").toString()
        databaseRef.addValueEventListener(object : ValueEventListener {
            override fun onDataChange(snapshot: DataSnapshot) {
                allOfferList.clear()
                for (dataSnapshot in snapshot.children) {
                    dataSnapshot.getValue(Offer::class.java)?.let {
                        if (it.buyer_name == businessName) {
                            allOfferList.add(it)
                        }
                    }
                }
                updateView()
            }

            override fun onCancelled(error: DatabaseError) {}
        })
    }

    private fun updateView() {
        if (allOfferList.size > 0) {
            binding.statsHolder.visibility = View.VISIBLE
            setUpMyOffers(allOfferList)
        } else {
            binding.offersEmpty.visibility = View.VISIBLE
            binding.imgOffersEmpty.visibility = View.VISIBLE
            binding.dataLoading.visibility = View.GONE
            binding.totalAccepted.text = "0"
            binding.totalPending.text = "0"
            binding.totalDeclined.text = "0"
        }
    }

    private fun setUpMyOffers(offers: List<Offer?>?) {
        val linearLayoutManager = LinearLayoutManager(activity, RecyclerView.VERTICAL, false)
        val offersAdapter = BuyerOffersAdapter(activity, offers)
        setUpStats(offers)
        binding.allOffers.adapter = offersAdapter
        binding.allOffers.layoutManager = linearLayoutManager
        binding.allOffers.visibility = View.VISIBLE
        binding.allOffers.visibility = View.VISIBLE
        binding.itemCard.visibility = View.VISIBLE
        binding.dataLoading.visibility = View.GONE
    }

    private fun setUpStats(offers: List<Offer?>?) {
        var accepted = 0
        var pending = 0
        var declined = 0
        for (i in offers!!.indices) {
            val o = offers[i]
            if (o!!.getOffer_status() == "Accepted") {
                accepted += 1
            } else if (o.getOffer_status() == "Pending") {
                pending += 1
            } else {
                declined += 1
            }
        }
        binding.totalAccepted.text = accepted.toString()
        binding.totalPending.text = pending.toString()
        binding.totalDeclined.text = declined.toString()
    }
}